crypto/blake2/blake2b.c

```diff
@@ -0,0 +1,310 @@
+    memset(P->reserved, 0, sizeof(P->reserved));
+    memset(P->salt,     0, sizeof(P->salt));
+    memset(P->personal, 0, sizeof(P->personal));
+    return blake2b_init_param(c, P);
+}
+
+int BLAKE2b_InitKey(BLAKE2B_CTX *c, const void *key, size_t keylen) {
+    BLAKE2B_PARAM P[1];
+    uint8_t block[BLAKE2B_BLOCKBYTES];
+
+    if (!key || !keylen || keylen > BLAKE2B_KEYBYTES) {
+        return 0;
+    }
+
+    P->digest_length = BLAKE2B_DIGEST_LENGTH;
+    P->key_length    = keylen;
+    P->fanout        = 1;
+    P->depth         = 1;
+    store32(&P->leaf_length, 0);
+    store64(&P->node_offset, 0);
+    P->node_depth    = 0;
+    P->inner_length  = 0;
+    memset(P->reserved, 0, sizeof(P->reserved));
+    memset(P->salt,     0, sizeof(P->salt));
+    memset(P->personal, 0, sizeof(P->personal));
+
+    if(!blake2b_init_param(c, P)) {
+        return 0;
+    }
+
+    memset(block, 0, BLAKE2B_BLOCKBYTES);
+    memcpy(block, key, keylen);
+    BLAKE2b_Update(c, block, BLAKE2B_BLOCKBYTES);
+    OPENSSL_cleanse(block, BLAKE2B_BLOCKBYTES); /* Burn the key from stack */
+    return 1;
+}
+
+static void blake2b_compress(BLAKE2B_CTX *S,
+                            const uint8_t block[BLAKE2B_BLOCKBYTES])
+{
+    uint64_t m[16];
+    uint64_t v[16];
+    int i;
+
+    for(i = 0; i < 16; ++i) {
+        m[i] = load64(block + i * sizeof(m[i]));
+    }
+
+    for(i = 0; i < 8; ++i) {
+        v[i] = S->h[i];
+    }
+
+    v[8]  = blake2b_IV[0];
+    v[9]  = blake2b_IV[1];
+    v[10] = blake2b_IV[2];
+    v[11] = blake2b_IV[3];
+    v[12] = S->t[0] ^ blake2b_IV[4];
+    v[13] = S->t[1] ^ blake2b_IV[5];
+    v[14] = S->f[0] ^ blake2b_IV[6];
+    v[15] = S->f[1] ^ blake2b_IV[7];
+#define G(r,i,a,b,c,d) \
+    do { \
+        a = a + b + m[blake2b_sigma[r][2*i+0]]; \
+        d = rotr64(d ^ a, 32); \
+        c = c + d; \
+        b = rotr64(b ^ c, 24); \
+        a = a + b + m[blake2b_sigma[r][2*i+1]]; \
+        d = rotr64(d ^ a, 16); \
+        c = c + d; \
+        b = rotr64(b ^ c, 63); \
+    } while(0)
+#define ROUND(r)  \
+    do { \
+        G(r,0,v[ 0],v[ 4],v[ 8],v[12]); \
+        G(r,1,v[ 1],v[ 5],v[ 9],v[13]); \
+        G(r,2,v[ 2],v[ 6],v[10],v[14]); \
+        G(r,3,v[ 3],v[ 7],v[11],v[15]); \
+        G(r,4,v[ 0],v[ 5],v[10],v[15]); \
+        G(r,5,v[ 1],v[ 6],v[11],v[12]); \
+        G(r,6,v[ 2],v[ 7],v[ 8],v[13]); \
+        G(r,7,v[ 3],v[ 4],v[ 9],v[14]); \
+    } while(0)
+    ROUND(0);
+    ROUND(1);
+    ROUND(2);
+    ROUND(3);
+    ROUND(4);
+    ROUND(5);
+    ROUND(6);
+    ROUND(7);
+    ROUND(8);
+    ROUND(9);
+    ROUND(10);
+    ROUND(11);
+
+    for(i = 0; i < 8; ++i) {
+        S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
+    }
+
+#undef G
+#undef ROUND
+}
+
+int BLAKE2b_Update(BLAKE2B_CTX *c, const void *data, size_t datalen) {
+    const uint8_t *in = data;
+    size_t left, fill;
+
+    while(datalen > 0) {
+        left = c->buflen;
+        fill = 2 * BLAKE2B_BLOCKBYTES - left;
+        if(datalen > fill) {
```

Comment:
On Sat, Mar 5, 2016 at 12:53 PM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55127766>:
>
> > +    for(i = 0; i < 8; ++i) {
> > +        S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
> > +    }
> > +
> > +#undef G
> > +#undef ROUND
> > +}
> > +
> > +int BLAKE2b_Update(BLAKE2B_CTX *c, const void *data, size_t datalen) {
> > +    const uint8_t *in = data;
> > +    size_t left, fill;
> > +
> > +    while(datalen > 0) {
> > +        left = c->buflen;
> > +        fill = 2 * BLAKE2B_BLOCKBYTES - left;
> > +        if(datalen > fill) {
>
> This should be >=
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/566/files#r55127766>.
>

No, > is correct.  It will actually cause the result to be incorrect if it
were changed.  BLAKE2b is as "lazy" as possible about compression.  The
last call to Update is different, so it matters whether or not this is >=
vs >.